src/opnsense/scripts/OPNsense/CaptivePortal/cp-background-process.py

```diff
@@ -50,18 +51,57 @@ def __init__(self):
         # open syslog and notice startup
         syslog.openlog('captiveportal', facility=syslog.LOG_LOCAL4)
         syslog.syslog(syslog.LOG_NOTICE, 'starting captiveportal background process')
-        # handles to ipfw, arp the config and the internal administration
-        self.ipfw = IPFW()
+        # handles to pf, arp, the config and the internal administration
         self.arp = ARP()
         self.cnf = Config()
         self.db = DB()
         self._conf_zone_info = self.cnf.get_zones()
+        self._accounting = dict()
+
 
     def list_zone_ids(self):
         """ return zone numbers
         """
         return self._conf_zone_info.keys()
 
+    def get_accounting(self):
+        for zoneid in self.list_zone_ids():
+            pf_stats = PF.list_accounting_info(zoneid)
+            current_clients = self.db.list_clients(zoneid)
+            for client in current_clients:
+                ip = client['ipAddress']
+                if ip not in pf_stats:
+                    if ip in self._accounting:
+                        # client removed from table, delete accounting as well
+                        del self._accounting[ip]
+                    continue
+
+                values = " ".join(map(str, [pf_stats[ip]['in_pkts'], pf_stats[ip]['in_bytes'],
+                                pf_stats[ip]['out_pkts'], pf_stats[ip]['out_bytes']]))
+                cur_hash = hashlib.md5(values.encode()).hexdigest()
+
+                if ip not in self._accounting:
+                    # initial accounting update if client is not known in memory
+                    self._accounting[ip] = {
+                        'last_accessed': client['last_accessed'],
+                        'last_accessed_hash': cur_hash,
+                        'in_pkts': 0,
+                        'in_bytes': 0,
+                        'out_pkts': 0,
+                        'out_bytes': 0
+                    }
+
+                    # merge counters as they are known in pf, last_accessed is
+                    # determined on next update to save us the work of querying the db stats
+                else:
```

Comment:
we likely want to exclude all zero's here, a reload flushes the counters, which doesn't change access time. Had to read all of this a couple of times, but given the boundaries that exist, this does seem like a nice trick to offer a decent approximation of last accessibility.

Reply:
@AdSchellevis I updated the logic such that the decision for an accounting sync is now also determined in the same function, based on the state of the db (this state is always correct after a zone sync). This means I moved the accounting update after the zone sync.

If an accounting sync is necessary, all counters are flushed as you mentioned, therefore we can provide zeroes for the initial addition in memory and bootstrap it with the last known access time from the db (the next iteration will pick up on last access change). Since we now sync the accounting (insert match rules) after the accounting update to the db, the counters will reset immediately after collecting the last numbers which serves to minimize the amount of data we miss.